cli/cmd/top.go

```diff
@@ -0,0 +1,372 @@
+	cmd.PersistentFlags().StringVar(&options.authority, "authority", options.authority,
+		"Display requests with this :authority")
+	cmd.PersistentFlags().StringVar(&options.path, "path", options.path,
+		"Display requests with paths that start with this prefix")
+
+	return cmd
+}
+
+func getTrafficByResourceFromAPI(w io.Writer, client pb.ApiClient, req *pb.TapByResourceRequest) error {
+
+	rsp, err := client.TapByResource(context.Background(), req)
+	if err != nil {
+		return err
+	}
+
+	err = termbox.Init()
+	if err != nil {
+		return err
+	}
+	defer termbox.Close()
+
+	requestCh := make(chan topRequest, 100)
+	done := make(chan struct{})
+
+	go recvEvents(rsp, requestCh, done)
+	go pollInput(done)
+
+	renderTable(requestCh, done)
+
+	return nil
+}
+
+func recvEvents(tapClient pb.Api_TapByResourceClient, requestCh chan<- topRequest, done chan<- struct{}) {
+	outstandingRequests := make(map[pb.TapEvent_Http_StreamId]topRequest)
+	for {
+		event, err := tapClient.Recv()
+		if err == io.EOF {
+			fmt.Println("Tap stream terminated")
+			close(done)
+			return
+		}
+		if err != nil {
+			fmt.Println(err.Error())
+			close(done)
+			return
+		}
+
+		switch ev := event.GetHttp().GetEvent().(type) {
+		case *pb.TapEvent_Http_RequestInit_:
+			id := *ev.RequestInit.GetId()
+			outstandingRequests[id] = topRequest{
+				event:   event,
+				reqInit: ev.RequestInit,
+			}
+
+		case *pb.TapEvent_Http_ResponseInit_:
+			id := *ev.ResponseInit.GetId()
+			if req, ok := outstandingRequests[id]; ok {
+				req.rspInit = ev.ResponseInit
+			} else {
+				log.Warn("Got ResponseInit for unknown stream: %d:%d", id.GetBase(), id.GetStream())
+			}
+
+		case *pb.TapEvent_Http_ResponseEnd_:
+			id := *ev.ResponseEnd.GetId()
+			if req, ok := outstandingRequests[id]; ok {
+				req.rspEnd = ev.ResponseEnd
+				requestCh <- req
+			} else {
+				log.Warn("Got ResponseEnd for unknown stream: %d:%d", id.GetBase(), id.GetStream())
+			}
+		}
+	}
+}
+
+func pollInput(done chan<- struct{}) {
+	for {
+		switch ev := termbox.PollEvent(); ev.Type {
+		case termbox.EventKey:
+			if ev.Ch == 'q' || ev.Key == termbox.KeyCtrlC {
+				close(done)
+				return
+			}
+		}
+	}
+}
+
+func renderTable(requestCh <-chan topRequest, done <-chan struct{}) {
+	ticker := time.NewTicker(100 * time.Millisecond)
+	var table []tableRow
+
+	for {
+		select {
+		case <-done:
+			return
+		case req := <-requestCh:
+			tableInsert(&table, req)
+		case <-ticker.C:
+			termbox.Clear(termbox.ColorDefault, termbox.ColorDefault)
+			renderHeaders()
+			renderTableBody(table)
+			termbox.Flush()
+		}
+	}
+}
+
+func tableInsert(table *[]tableRow, req topRequest) {
+
+	by := req.reqInit.GetPath()
+	source := stripPort(addr.PublicAddressToString(req.event.GetSource()))
+	destination := stripPort(addr.PublicAddressToString(req.event.GetDestination()))
+	if pod := req.event.DestinationMeta.Labels["pod"]; pod != "" {
+		destination = pod
+	}
+	latency, latencyErr := ptypes.Duration(req.rspEnd.GetSinceRequestInit())
```

Comment:
It feels awkward to not check the error from this call immediately, and instead only check it when updating the row in the loop below.

I think realistically, we would never expect for this call to result in an error. If it does, maybe we should reject the whole row as invalid. Something like:

```
	latency, err := ptypes.Duration(req.rspEnd.GetSinceRequestInit())
	if err != nil {
		log.Errorf("error parsing duration %v: %s", req.rspEnd.GetSinceRequestInit(), err)
		return
	}
```

Or if we're not concerned about some requests having 0 latency, then you could just log the error and not return early:

```
	latency, err := ptypes.Duration(req.rspEnd.GetSinceRequestInit())
	if err != nil {
		log.Errorf("error parsing duration %v: %s", req.rspEnd.GetSinceRequestInit(), err)
	}
```

But either way, you should try to check the error immediately if possible.